### ISSUE TITLE
Some font kerning/hinting not properly handled and justified

### ISSUE THREAD

**Issue by Omikhleia** [sile-typesetter/sile#1358]
SILE 0.12.4

I had this bug for a long time at lots of place in my ~350-page test book, but had failed to get a decent test so far... Whee, I got it at last on a minimal test (= no French, no hyphenation occurring, no fi or ff ligatures etc.)

It's clearly related to the font (= LibertinusSerif-Regular.otf, version 7.040), but I do not observe such issue with it in, say, LibreOffice (toying with OpenType features too, see analysis below), so I am suspecting it could also be due to SILE.

Test script:
```
\begin[class=book, papersize=6in x 9in]{document}
\font[family=Libertinus Serif, size=11pt]
\neverindent

Lorem ipsum dolor, sit amet consetetur, sadipscing elitr, sed diam non umyeirmod tempor invidunt ut labore et dolore magna aliquyam erat sed diam
voluptua at vero eos et accusam et justo duo dolores et ea rebum stet clita
kasd

Lorem ipsum dolor{}, sit amet consetetur{}, sadipscing elitr{}, sed diam non umyeirmod tempor invidunt ut labore et dolore magna aliquyam erat sed diam
voluptua at vero eos et accusam et justo duo dolores et ea rebum stet clita
kasd

\end{document}
```

Expected result: both paragraphs are justified properly.

Observed result: first paragraph is not justified properly, there's some off space at the end of some lines, as shown below:

![image](https://user-images.githubusercontent.com/18075640/162944478-6d739604-8631-493b-af48-be5abd10f524.png)

It is slight, but visible (and on some real text lines, _very_ visible by naked eye on a printed page).

(Pre-)Analysis:

This font seems to have some special kerning (or is that "hinting"?) after some characters, here the **r**, at least when followed by punctuation (same issue, for instance with a dot instead of the comma) which fail to be taken into account in line width computation. Manually cancelling it (inserting a `{}` as illustrated on the second paragraph) shows correct justification.

Workaround (besides the `{}` trick)

```
\font[family=Libertinus Serif, size=11pt, features="-kern"]
```

I am a bit sad to disable that font feature, though. Some words look better with it (e.g. "Wait") without having, it seems so far, the same consequence.

**Comment by alerque** [comment]
Clearly a bug, probably in SILE (although if I wasn't also the font maintainer as well it would be oh-so-convenient to just label this 'not my bug').

Judging so far just from your screen short, it looks like SILE might be *over* applying (possible multiple times?) the relavant kern values. Consider this:

![image](https://user-images.githubusercontent.com/173595/162964569-189482a7-5c9e-4205-919d-72de073a7757.png)

**Comment by Omikhleia** [comment]
Modifying `languages/unicode.lua` as follows:
```
    handleWordBreak = function (self, item)
      -- self:makeToken() -- <---------------- REMOVED
      if self:isSpace(item.text) then
        -- Spacing word break
        self:makeToken() -- <-------------- MOVED HERE (so only for a space glue)
        self:makeGlue(item)
      else -- a word break which isn't a space
        -- print("ICU word break not a space", item.text) -- <-- If need be to check the break item
        self:addToken(item.text, item)
      end
    end,
```

Then the problem "seems" solved (at least, until other issues happen later?).

It wouldn't be illogical: the ICU break here is triggered on the comma (I checked by printing `item.text`, with the above test script), and the current code splits it from the previous tokens. I already had some doubts, when working on the French typography, why we have two different types of line/word-breaking logic, with the Knuth line-breaking and the ICU unicode parsing. I don't understand all of it very well here, but it would seem that at some point we have conflicting approaches, so the tokens are probably considered at one point as being split (so the negative font kern between the r and the punctuation is not applied) and at another as unsplit (so the kerning is applied).

**Comment by Omikhleia** [comment]
As far as I can tell, #1378 also fixes this issue. I was reluctant initially to consider this was the same as #1362 (since the proposed change to `handleWordBreak` above improved this one and it henceforth seemed to be of a somewhat different nature), but maybe it is after all, for reasons that elude my current understanding.

**Comment by Omikhleia** [comment]
As stated above, since #1378 landed in 0.13.0, this case seems closed too.
About 1 year later, I have not met it again in my projects -- So I am closing it. Though a "workaround" of sorts, #1378 did the necessary tricks, and if something similar even pops up in the future, it will need to be addressed in that new context.
